### PR TITLE
ENH PHP 8.2 support

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/10403

Installer run - https://github.com/creative-commoners/silverstripe-installer/actions/runs/3745680107

### CODE REVIEW

File: src/ORM/Connect/MySQLStatement.php [hunk 1]
```diff
@@ -73,7 +73,6 @@ public function __construct($statement, $metadata)
     public function __destruct()
     {
         $this->statement->close();
-        $this->currentRecord = false;
```
Comment (emteknetnz) [inline src/ORM/Connect/MySQLStatement.php:76]: This property is totally unused so not even adding it as a class property

File: composer.json [hunk 2]
```diff
@@ -27,7 +27,7 @@
         "composer/installers": "^2.2",
         "guzzlehttp/guzzle": "^7.5.0",
         "guzzlehttp/psr7": "^2.4.0",
-        "embed/embed": "^4.4.4",
+        "embed/embed": "^4.4.7",
```
Comment (emteknetnz) [inline composer.json:30]: This is required to get this fix https://github.com/oscarotero/Embed/commit/7c4c38be6826a34f9b9508e5449a5fcb01bf515e

File: src/Control/HTTPRequest.php [hunk 3]
```diff
@@ -579,7 +579,7 @@ public function match($pattern, $shiftOnSuccess = false)
                         $shiftCount = sizeof($patternParts ?? []);
                         $remaining = count($this->dirParts ?? []) - $i;
                         for ($j = 1; $j <= $remaining; $j++) {
-                            $arguments["$${j}"] = $this->dirParts[$j + $i - 1];
+                            $arguments['$' . $j] = $this->dirParts[$j + $i - 1];
```
Comment (emteknetnz) [inline src/Control/HTTPRequest.php:582]: Have confirmed with local testing that this is the correct way to de-deprecate this extremely confusing syntax

File: src/ORM/Connect/MySQLDatabase.php [hunk 4]
```diff
@@ -65,6 +65,8 @@ class MySQLDatabase extends Database implements TransactionManager
      */
     private $transactionManager = null;
 
+    private int $transactionNesting = 0;
```
Comment (emteknetnz) [inline src/ORM/Connect/MySQLDatabase.php:68]: This is to fix a dynamic property property elsewhere in the file i.e. `$this->transactionNesting = ...`

Searching for `->transactionNesting` in the core codebase only yields `$this->transationNesting`, and only only classes that themselves define the `transationNesting` property, so no need to make this anything other than `private`

File: thirdparty/php-peg/Parser.php [hunk 5]
```diff
@@ -9,6 +9,13 @@
  *  the bracket if a failed match + restore has moved the current position backwards - so we have to check that too.
  */
 class ParserRegexp {
+
+	public $parser;
+	public $rx;
+	public $matches;
+	public $match_pos;
+	public $check_pos;
```
Comment (emteknetnz) [inline thirdparty/php-peg/Parser.php:17]: These were all dynamic properties - declaring them as un-typed public properties makes them functionally identical

File: src/View/ViewableData.php [hunk 6]
```diff
@@ -28,6 +28,7 @@
  * is provided and automatically escaped by ViewableData. Any class that needs to be available to a view (controllers,
  * {@link DataObject}s, page controls) should inherit from this class.
  */
+#[\AllowDynamicProperties]
```
Comment (GuySartorelli) [inline src/View/ViewableData.php]: What specifically is this resolving that we can't resolve by using non-deprecated functionality?
Adding this is effectively kicking the can down the road for when they remove this attribute.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/View/ViewableData.php (hunk 6 above)
Comment (emteknetnz) [reply src/View/ViewableData.php]: Apologies, forgot to put explainer on this one

It's for __set() which calls `setField()` which calls `$this->$field = $value;`

`getField()` and `hasField()` both also use `$this->$field`

I was worried about the consequences of changing this since ViewableData is such a base class, though I'll update the PR to use `$this->data[$field]` instead and see if anything breaks

File: src/Dev/Constraint/SSListContains.php [hunk 7]
```diff
@@ -20,6 +20,8 @@ class SSListContains extends Constraint implements TestOnly
      */
     protected $matches = [];
 
+    protected SSListExporter $exporter;
```
Comment (emteknetnz) [inline src/Dev/Constraint/SSListContains.php:23]: This was being set as a dynamic property in __construct()

File: src/Dev/Constraint/SSListContainsOnlyMatchingItems.php [hunk 8]
```diff
@@ -19,6 +19,8 @@ class SSListContainsOnlyMatchingItems extends Constraint implements TestOnly
      */
     private $match;
 
+    protected SSListExporter $exporter;
```
Comment (emteknetnz) [inline src/Dev/Constraint/SSListContainsOnlyMatchingItems.php:22]: This was being set as a dynamic property in __construct()

File: src/Dev/SapphireTest.php [hunk 9]
```diff
@@ -161,6 +161,8 @@ abstract class SapphireTest extends TestCase implements TestOnly
      */
     protected static $tempDB = null;
 
+    protected FixtureFactory|bool $fixtureFactory;
```
Comment (emteknetnz) [inline src/Dev/SapphireTest.php:164]: This was being set as a dynamic property in setUp() via `$fixtureState->getFixtureFactory(static::class);` - getFixtureFactory() can return `FixtureFactory|bool` hence why I've used that as the type.

File: src/View/ViewableData.php [hunk 10]
```diff
@@ -66,6 +66,11 @@ class ViewableData implements IteratorAggregate
      */
     private static $casting_cache = [];
 
+    /**
+     * Acts as a PHP 8.2+ compliant replacement for dynamic properties
+     */
+    private array $data = [];
```
Comment (emteknetnz) [inline src/View/ViewableData.php:72]: ViewableData is a base class and there's simply far too many magic calls in Silverstripe to full stop using dynamic properties.  Realistically it's either a `private array` or just add `#[\AllowDynamicProperties]` to the class

File: src/Dev/FixtureBlueprint.php [hunk 11]
```diff
@@ -70,6 +72,17 @@ public function __construct($name, $class = null, $defaults = [])
         $this->defaults = $defaults;
     }
 
+    public function getFactory(): FixtureFactory
```
Comment (emteknetnz) [inline src/Dev/FixtureBlueprint.php:75]: There's a private static $dependencies that includes 'factory' - the get/set with a private property is a pretty standard way to handle this in the reset of the codebase.  Previously this was done via a dynamic property

File: tests/php/ORM/DataObjectTest.php [hunk 12]
```diff
@@ -826,10 +826,6 @@ public function testHasOneAsField()
         $team1->Captain = $captain2;
         $team1->write();
         $this->assertEquals($captain2->ID, $team1->Captain->ID);
-
-        // Setter: Custom data (required by DataDifferencer)
```
Comment (emteknetnz) [inline tests/php/ORM/DataObjectTest.php:830]: I looked into this and experimented with some test objects on my local with $has_one relationships and putting them into DataDifferencer - as far as I can tell this commet about it being required by DataDifferencer is incorrect.  The existing DataDifferenceTest still passes its unit tests and does not throw deprecation warnings.

I deleted this unit test because it end up trying to create a dynamic property called Captain in (from memory)  DBHTMLText and a deprecation warning was thrown

File: src/ORM/DataObject.php [hunk 13]
```diff
@@ -2834,55 +2834,52 @@ public function setField($fieldName, $val)
 
         // Situation 1: Passing an DBField
         if ($val instanceof DBField) {
-            $val->setName($fieldName);
-            $val->saveInto($this);
```
Comment (emteknetnz) [inline src/ORM/DataObject.php]: Note: this change has caused a [bunch of new failures](https://github.com/silverstripe/silverstripe-framework/actions/runs/3708565574/jobs/6286263995#step:12:116), so I may well revert it. CI was green prior to this change, though there were dynamic property deprecation warnings

It looks like a lot has changed in this file though it hasn't - previously it was at a high level:
```php
if ($val instanceof DBField) {
  // handle DBField
  $val->saveInto($this);
} else {
  // handle non-DBField $val
  $this->record[$fieldName] = $val;
}
```

Now it is

```php
if ($val instanceof DBField) {
  $val = $val->getValue()
}
// handle non-DBField $val
$this->record[$fieldName] = $val;
```

The reason this was done was was to fix a dynamic property deprecation warning caused by SecurityTest::testExternalBackUrlRedirectionDisallowed() where at attempt was made to do a magic-method __set() using a DBField as a value.

The deprecation warning was thrown from [here](https://github.com/silverstripe/silverstripe-framework/blob/4.12/src/ORM/FieldType/DBField.php#L545)

The order of calls made is:
```
ChangePasswordHandler::doChangePassword()
  $member->AutoLoginExpired = DBDatetime::create()->now(); << value is set as a DBField
ViewableData::__set() <<< first call to __set()
DataObject::setField()
DBField::saveInto()
  $dataObject->$fieldName = $this->value; <<< can't re-call __set(), dynamic property deprecation warning
```

File: src/ORM/DataObject.php [hunk 14]
```diff
@@ -2841,49 +2842,62 @@ public function setField($fieldName, $val)
...
-                }
-            }
+    /**
+     * Usually you should call setField() instead
+     *
+     * Intended to called from setField(), or DBField::saveInto() which is called from setField()
+     * that used to use dynamic properties e.g. $dataObject->$fieldName = $value
+     * prior to dynamic properties being deprecationed in PHP 8.2
+     */
+    public function setFieldValue(string $fieldName, mixed $val): void
```
Comment (emteknetnz) [inline src/ORM/DataObject.php]: Logic for this method was cut and pasted from setField()

Has the added benefit that when a DBField is passed as $val in setField(), $fieldName is now added to $this->changed whereas previously it was (probably incorrectly) not

File: src/ORM/DataObject.php [hunk 15]
```diff
@@ -2841,49 +2842,62 @@ public function setField($fieldName, $val)
...
+    {
+        $schema = static::getSchema();
+        // If this is a proper database field, we shouldn't be getting non-DBField objects
+        if (is_object($val) && !($val instanceof DBField) && $schema->fieldSpec(static::class, $fieldName)) {
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php:2858]: Where did the `!($val instanceof DBField)` check come from? That doesn't seem related to making this code PHP 8.2 compatible? What is the point of the `$schema->fieldSpec(static::class, $fieldName)` check here, if we're changing the condition?

File: src/ORM/DataObject.php [hunk 16]
```diff
@@ -2841,49 +2842,62 @@ public function setField($fieldName, $val)
...
-                        )
-                    );
-                }
-            }
+    /**
+     * Usually you should call setField() instead
+     *
+     * Intended to called from setField(), or DBField::saveInto() which is called from setField()
+     * that used to use dynamic properties e.g. $dataObject->$fieldName = $value
+     * prior to dynamic properties being deprecationed in PHP 8.2
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]:
```suggestion
```
I don't think that context is useful for developers.

File: src/ORM/DataObject.php [hunk 17]
```diff
@@ -2841,49 +2842,62 @@ public function setField($fieldName, $val)
             // later referenced to update the parent dataobject
             if ($val instanceof DBComposite) {
                 $val->bindTo($this);
-                $this->record[$fieldName] = $val;
+                $this->setFieldValue($fieldName, $val);
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php:2845]: How does this related to PHP 8.2 compatibility and what is the intention of this change? It seems like we're now doing a whole load of processing that wasn't being done before?

If this is just for setting the changed logic, it would probably be better to instead have a `protected function markIfChanged(string $fieldName, mixed $value)` instead, since that seems to be the only part of the logic that needs to be duplicated.

The typing checks that are currently in `setFieldValue()` could then remain in the `else` block here, and the API will be a little clearer (having public `setField()` and `setFieldValue()` is a bit confusing even if the phpdoc does clarify the purpose)

File: src/ORM/FieldType/DBBoolean.php [hunk 18]
```diff
@@ -46,7 +46,7 @@ public function saveInto($dataObject)
     {
         $fieldName = $this->name;
         if ($fieldName) {
-            $dataObject->$fieldName = ($this->value) ? 1 : 0;
+            $dataObject->setFieldValue($fieldName, $this->value ? 1 : 0);
```
Comment (GuySartorelli) [inline src/ORM/FieldType/DBBoolean.php]: `$dataObject->$fieldName = $someValue` will go through `__set()` which will then call `setField()` which ultimately has the `setFieldValue()` logic in it - I don't see any reason to subvert that logic flow, especially as it then means any `setX()` method will _not_ be called.
Can you please clarify what problem this change is solving? I may be missing some context here.

File: src/View/ViewableData_Customised.php [hunk 19]
```diff
@@ -59,19 +59,26 @@ public function hasMethod($method)
...
-        if ($this->customised->hasField($fieldName) || $this->customised->hasMethod($fieldName)) {
+        if ($this->customisedHas($fieldName)) {
             return $this->customised->obj($fieldName, $arguments, $cache, $cacheName);
         }
         return $this->original->obj($fieldName, $arguments, $cache, $cacheName);
     }
+
+    private function customisedHas(string $fieldName): bool
+    {
+        return property_exists($this->customised, $fieldName) ||
```
Comment (GuySartorelli) [inline src/View/ViewableData_Customised.php:80]: Why explicitly check `property_exists` here? Doesn't `hasField()` check that already?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/ORM/FieldType/DBBoolean.php (hunk 18 above)
Comment (emteknetnz) [reply src/ORM/FieldType/DBBoolean.php]: Once you are inside `__set()` - in this case ViewableData::__set(), any futher call to `$dataObject->$fieldName` won't call `__set()` again as this would cause an infinite loop.  Instead at this point it's setting properties directly, which in most causes will be setting dynamic properties which are deprecated, hence the code change.

Let's say I'm call $MyDataObject->MyField = DBBoolean::create('MyField'); from PageController::init().  The code path is now:

```
ViewableData::__set() .. $this->setField($property, $value);
DataObject::setField() .. if ($val instanceof DBField) ..  $val->saveInto($this);
DBBoolean::saveInto() .. $dataObject->setFieldValue($fieldName, $this->value ? 1 : 0);
DataObject::setFieldValue() .. $this->record[$fieldName] = $val;
```

File: src/View/ViewableData_Customised.php (hunk 19 above)
Comment (emteknetnz) [reply src/View/ViewableData_Customised.php:80]: `ViewableData::hasField()` does, but it can be overridden, for instance `DataObject::hasField()` does not because it's explicitly checking for database columns only and `get$field()`

File: src/ORM/DataObject.php (hunk 16 above)
Comment (emteknetnz) [reply src/ORM/DataObject.php]: Removed

File: src/ORM/DataObject.php (hunk 15 above)
Comment (emteknetnz) [reply src/ORM/DataObject.php:2858]: > Where did the !($val instanceof DBField) check come from?

Previously DBField wasn't going through this code path, now it is.

The important bit for PHP 8.2 is `$this->record[$fieldName] = $val;`.  `DataObject::setFieldValue() `is called by implementations of `DBField::saveInto()` which used to call `$dataObject->$fieldName = $val;` which they can no longer do.

File: src/ORM/DataObject.php (hunk 17 above)
Comment (emteknetnz) [reply src/ORM/DataObject.php:2845]: > It seems like we're now doing a whole load of processing that wasn't being done before?

The extra processing is just to (correctly I think) update the "changed" status when $val is a DBField.  The extra processing is a few `array_key_exists()` type of checks which will have a negligible performance impact so I'm not worried about it at all.  Those checks already happen for non DBField values.

`protected function markIfChanged(string $fieldName, mixed $value)` - that's just adding extra API when we already have `setFieldValue()`? Note we do need `setFieldValue()` for the various DBField implementations of `saveInto()`, which I've talked about below in the comment about `!($val instanceof DBField)`

To do what you're proposing above, I think we'd need to remove the `saveInto()` methods and consolidate everything in `DataObject::setField()` (i could be totally wrong though)

File: src/View/ViewableData_Customised.php (hunk 19 above)
Comment (GuySartorelli) [reply src/View/ViewableData_Customised.php:80]: `DataObject::hasField()` is irrelevant, since `DataObject` is not a subclass of `ViewableData_Customised`... that said, I _suppose_ someone could inject a custom copy of `ViewableData_Customised` somewhere which doesn't include the `property_exists` check in `hasField()`. Unlikely, but this protects against it with no downside that I can see so I'm okay with it.

File: src/ORM/FieldType/DBBoolean.php (hunk 18 above)
Comment (GuySartorelli) [reply src/ORM/FieldType/DBBoolean.php]: Ahh okay, I was missing the context that it wouldn't call `__set()` again.

Can we not just call `setField()` here then instead of `setFieldValue()`? We don't need an entirely new public API to do this.

Comment (emteknetnz) [reply src/ORM/FieldType/DBBoolean.php]: No ... see the codepath above, at this point were are in `DBBoolean::saveInto()`, if we call `DataObject::setField()` we'll go into an infinite loop

File: src/View/ViewableData_Customised.php (hunk 19 above)
Comment (emteknetnz) [reply src/View/ViewableData_Customised.php:80]: the `property_exists()` check is on `$this->customised`, not `$this`.  `$this->customed` is a `ViewableData`, NOT a `ViewableData_Customised`

Despite the fact that `ViewableData_Customised` extends `ViewableData`, it is in fact a container for a pair of ViewableData's - `$this->original` and `$this->customised`

File: src/ORM/FieldType/DBBoolean.php (hunk 18 above)
Comment (GuySartorelli) [reply src/ORM/FieldType/DBBoolean.php]: We won't go into an infinite loop, because the value that `DBBoolean::saveInto()` passes in is _not_ an instance of `DBField` - I suppose technically it could be, but that would be very unusual, and in any case it won't be _itself_ so it will eventually resolve down to a value.
The path then becomes:

```
ViewableData::__set() .. $this->setField($property, $value);
DataObject::setField() .. if ($val instanceof DBField) ..  $val->saveInto($this);
DBBoolean::saveInto() .. $dataObject->setField($fieldName, $this->value ? 1 : 0);
DataObject::setField() .. $this->record[$fieldName] = $val;
```

Comment (emteknetnz) [reply src/ORM/FieldType/DBBoolean.php]: That would certainly be preferable to not have this new API, though I'm not sure that we can do this easily due to DBComposite.  There is unit testing to validate that you can actually set a dynamic property as a DBField currently. 
e.g. 
https://github.com/silverstripe/silverstripe-framework/blob/4/tests/php/ORM/DBCompositeTest.php#L33

There is this existing logic specifically for DBComposite in DataObject::setField()
```php
            // Situation 1a: Composite fields should remain bound in case they are
            // later referenced to update the parent dataobject
            if ($val instanceof DBComposite) {
                $val->bindTo($this);
                $this->record[$fieldName] = $val; // <<<<<< this
            }
```

Comment (GuySartorelli) [reply src/ORM/FieldType/DBBoolean.php]: I don't understand the problem.... the test that you've pointed out passes if you call `setField()` instead of `setFieldValue()`. I can't see any problems with calling `setField()` where you're currently calling `setFieldValue()` in the various `DBField` classes.

Comment (emteknetnz) [reply src/ORM/FieldType/DBBoolean.php]: Oh right, because we're retaining the line I just highlighted :laughing:

**Review by GuySartorelli — APPROVED** [review]
